src/Tools/dotnet-monitor/Extensibility/ProgramExtension.cs

```diff
@@ -0,0 +1,96 @@
+        public string DisplayName => Path.GetDirectoryName(_path);
+
+        /// <inheritdoc/>
+        bool IExtension.TryGetTypedExtension<TExtensionType>(out TExtensionType extension)
+        {
+            if (typeof(TExtensionType) == typeof(IEgressExtension))
+            {
+                extension = (TExtensionType) (IExtension) this;
+                return true;
+            }
+
+            extension = null;
+            return false;
+        }
+
+        /// <inheritdoc/>
+        public async Task<EgressArtifactResult> EgressArtifact(ExtensionEgressPayload configPayload, Func<Stream, CancellationToken, Task> getStreamAction, CancellationToken token)
+        {
+            // This _should_ only be used in this method, it can get moved to a constants class if that changes
+            const string CommandArgProviderName = "--Provider-Name";
+            // This is really weird, yes, but this is one of 2 overloads for [Stream].WriteAsync(...) that supports a CancellationToken, so we use a ReadOnlyMemory<char> instead of a string.
+            ReadOnlyMemory<char> NewLine = new ReadOnlyMemory<char>("\r\n".ToCharArray());
+
+            /* [TODOs]
+             * 1. [Done] Add a new service to dynamically find these extension(s)
+             * 2. [Done] Remove all raw logging statements from this method and refactor into LoggingExtensions
+             * 3. Stream StdOut and StdErr async in the process so their streams don't need to end before we can return
+             * 4. [Done] Refactor WaitForExit to do an async wait
+             * 5. Add well-factored protocol for returning information from an extension
+             */
+            ProcessStartInfo pStart = new ProcessStartInfo()
+            {
+                FileName = _path,
+                RedirectStandardInput = true,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                CreateNoWindow = true,
+                UseShellExecute = false,
+            };
+            pStart.ArgumentList.Add(ExtensionTypes.Egress);
+            pStart.ArgumentList.Add(CommandArgProviderName);
+            pStart.ArgumentList.Add(configPayload.ProviderName);
+
+            Process p = new Process()
+            {
+                StartInfo = pStart,
+            };
+
+            using OutputParser<EgressArtifactResult> parser = new(p, _logger);
+
+            _logger.ExtensionStarting(_path, pStart.Arguments);
+            p.Start();
+            await JsonSerializer.SerializeAsync<ExtensionEgressPayload>(p.StandardInput.BaseStream, configPayload, options: null, token);
+            await p.StandardInput.WriteAsync(NewLine, token);
+            await p.StandardInput.BaseStream.FlushAsync(token);
+            _logger.ExtensionConfigured(_path, p.Id);
+
+            await getStreamAction(p.StandardInput.BaseStream, token);
+            await p.StandardInput.BaseStream.FlushAsync(token);
+            p.StandardInput.Close();
```

Comment:
Is this necessary, especially after you've explicitly flushed the stream?

Reply:
Yep, this is necessary. This close is what notifies the extension that the stream is complete.